extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -538,33 +545,77 @@ export class AmpA4A extends AMP.BaseElement {
+   * @return {!Array<string>} A list of signing services.
+   */
+  getSigningServiceNames() {
+    // TODO(levitzky) Add dev key name once it goes live.
+    return getMode().localDev ? ['google'] : ['google'];
+  }
+
+  /**
+   * Retrieves all public keys, as specified in _a4a-config.js.
+   * None of the (inner or outer) promises returned by this function can reject.
+   *
+   * @return {!Array<!Promise<!Array<!Promise<?PublicKeyInfoDef>>>>}
    * @private
    */
-  validateAdResponse_(creative, signature) {
-    // Validate when we have a signature and we have native crypto.
-    if (!signature) {
-      // Guaranteed not a AMP creative.
-      return Promise.resolve(null);
+  getKeyInfoSets_() {
+    if (!isCryptoAvailable()) {
+      return [];
     }
-    if (verifySignatureIsAvailable()) {
-      // Among other things, the signature might not be proper base64.
-      // TODO(a4a-cam): This call used to be missing the conversion
-      // from ArrayBuffer to Uint8Array.  Strangely, that didn't cause
-      // any unit tests to fail, either locally or on Travis.  That
-      // indicates that the tests are too weak or aren't reporting
-      // correctly.  Check out and fix the tests.
-      return verifySignature(
-          new Uint8Array(creative), signature, publicKeyInfos).then(isValid => {
-            return isValid ? creative : null;
-          });
+    const jwkSetPromises = this.getSigningServiceNames().map(serviceName => {
+      // @param {string} serviceName
+      // @return {!Promise<!Array<!Object>>}
+      const url = signingServerURLs[serviceName];
+      if (url) {
+        return xhrFor(this.win).fetchJson(url, {mode: 'cors', method: 'GET'})
+            .then(jwkSetObj => {
+              // @param {*} jwkSetObj
+              // @return {!Promise<!Array<!Object>>}
+              if (isObject(jwkSetObj) && isArray(jwkSetObj.keys) &&
+                  jwkSetObj.keys.every(isObject)) {
+                return Promise.resolve(jwkSetObj.keys);
+              } else {
+                user().error(
+                    'Amp Ad',
+                    'Invalid response from signing server.',
+                    this.element);
+                return Promise.resolve([]);
+              }
+            }).catch(err => {
+              // @param {*} err
+              // @return {!Promise<!Array<*>>}
+              user().error('Amp Ad', err, this.element);
+              return Promise.resolve([]);
+            });
+      } else {
+        // The given serviceName does not have a corresponding URL in
+        // _a4a-config.js.
+        const reason = `Signing service '${serviceName}' does not exist.`;
+        user().error('Amp Ad', reason, this.element);
+        return Promise.resolve([]);
+      }
+    });
+    if (getMode().localDev) {
+      jwkSetPromises.push(Promise.resolve(devJwkSet));
     }
-    return Promise.reject('Public key validation of A4A ads not available');
+    return jwkSetPromises.map(jwkSetPromise =>
```

Comment:
I'm not worrying too much about the concurrency issue. I think such util is valuable to the whole code base. Instead of patch it to `Promise`, @cramforce suggested to make it a helper function, maybe under `src/utils/promises.js`.